cohortextractor/dsl.py

```diff
@@ -0,0 +1,185 @@
+PatientSeries or a single value to produce a new PatientSeries.
+
+All classes except Cohort are intended to be immutable.
+
+Methods are designed so that users can only perform actions that are semantically
+meaningful.  This means that the order of operations is restricted.  In a terrible ASCII
+railway diagram:
+
+
+             +---+
+             |   V
+      filter |  PatientFrame
+             |   |   |
+             +---+   | select_column
+                     V
+                PatientSeries
+
+
+             +---+
+             |   V
+      filter |  EventFrame -----------------------+
+             |   |   |                            |
+             +---+   | sort_by                    |
+                     V                            |
+             SortedEventFrame                     |
+                     |                            |
+        +---+        | (first/last)_for_patient   | (count/exists)_for_patient
+        |   V        V                            |
+ filter |  AggregatedEventFrame                   |
+        |   |        |                            |
+        +---+        | select_column              |
+                     V                            |
+               PatientSeries <--------------------+
+
+
+To support providing helpful error messages, we can implement __getattr__ on each class.
+This will intercept any lookup of a missing attribute, so that if eg a user tries to
+select a column from a SortedEventFrame, we can tell them they need to aggregate the
+SortedEventFrame first_for_patient.
+
+This docstring, and the function docstrings in this module are not currently intended
+for end users.
+"""
+
+from __future__ import annotations
+
+from cohortextractor.query_language import BaseTable as QMBaseTable
+from cohortextractor.query_language import Row
+from cohortextractor.query_language import Table as QMTable
+from cohortextractor.query_language import Value, ValueFromAggregate
+
+
+class Cohort:
+    """Represents the cohort of patients in a study."""
+
+    def set_population(self, population: PatientSeries) -> None:
+        """Set the population variable for this cohort."""
+
+        self.population = population
+        value = population.value
+
+        if not (isinstance(value, ValueFromAggregate) and value.function == "exists"):
+            raise ValueError(
+                "Population variable must return a boolean. Did you mean to use `exists_for_patient()`?"
+            )
+
+    def add_variable(self, name: str, variable: PatientSeries) -> None:
+        """Add a variable to this cohort by name."""
+
+        self.__setattr__(name, variable)
+
+    def __setattr__(self, name: str, variable: PatientSeries) -> None:
+        if not isinstance(variable, PatientSeries):
+            raise TypeError(
+                f"{name} must be a single value per patient (got '{variable.__class__.__name__}')"
+            )
+        self.__dict__[name] = variable
```

Comment:
Is there a reason for not spelling this `super().__setattr__(name, variable)`? To me that seems less invasive.

Reply:
I did this, and it was because `__setattr__` used to call `add_variable`.  I think `super` should be fine now

Reply:
I'll change.

Reply:
Addressed in 33af403bb3b7560b4abd43449fb407488160ad2b.